numba/dppl/tests/dppl/test_ufuncs.py

```diff
@@ -0,0 +1,1927 @@
+            (np.array([-0.5, 0.0, 0.5], dtype='f4'), types.Array(types.float32, 1, 'C')),
+            (np.array([-0.5, 0.0, 0.5], dtype='f8'), types.Array(types.float64, 1, 'C')),
+
+            (np.array([0,1], dtype=np.int8), types.Array(types.int8, 1, 'C')),
+            (np.array([0,1], dtype=np.int16), types.Array(types.int16, 1, 'C')),
+            (np.array([0,1], dtype=np.uint8), types.Array(types.uint8, 1, 'C')),
+            (np.array([0,1], dtype=np.uint16), types.Array(types.uint16, 1, 'C')),
+            ]
+
+    def _determine_output_type(self, input_type, int_output_type=None,
+                               float_output_type=None):
+        ty = input_type
+        if isinstance(ty, types.Array):
+            ty = ty.dtype
+
+        if ty in types.signed_domain:
+            if int_output_type:
+                output_type = types.Array(int_output_type, 1, 'C')
+            else:
+                output_type = types.Array(ty, 1, 'C')
+        elif ty in types.unsigned_domain:
+            if int_output_type:
+                output_type = types.Array(int_output_type, 1, 'C')
+            else:
+                output_type = types.Array(ty, 1, 'C')
+        else:
+            if float_output_type:
+                output_type = types.Array(float_output_type, 1, 'C')
+            else:
+                output_type = types.Array(ty, 1, 'C')
+        return output_type
+
+
+class TestUFuncs(BaseUFuncTest, TestCase):
+
+    def basic_ufunc_test(self, ufunc, flags=no_pyobj_flags,
+                         skip_inputs=[], additional_inputs=[],
+                         int_output_type=None, float_output_type=None,
+                         kinds='ifc', positive_only=False):
+
+        # Necessary to avoid some Numpy warnings being silenced, despite
+        # the simplefilter() call below.
+        self.reset_module_warnings(__name__)
+
+        pyfunc = _make_ufunc_usecase(ufunc)
+
+        inputs = list(self.inputs) + additional_inputs
+
+        for input_tuple in inputs:
+            input_operand = input_tuple[0]
+            input_type = input_tuple[1]
+
+            is_tuple = isinstance(input_operand, tuple)
+            if is_tuple:
+                args = input_operand
+            else:
+                args = (input_operand,) * ufunc.nin
+
+            if input_type in skip_inputs:
+                continue
+            if positive_only and np.any(args[0] < 0):
+                continue
+
+            # Some ufuncs don't allow all kinds of arguments
+            if (args[0].dtype.kind not in kinds):
+                continue
+
+            output_type = self._determine_output_type(
+                input_type, int_output_type, float_output_type)
+
+            input_types = (input_type,) * ufunc.nin
+            output_types = (output_type,) * ufunc.nout
+            cfunc_kernel = dppl.kernel(pyfunc)
+            cfunc = njit(parallel={'offload':True})(pyfunc)
```

Comment:
Unfortunately numba is not rewriting this ufuncs as array expression because the output argument is being provided. (https://github.com/IntelPython/numba/blob/pydppl/numba/np/ufunc/array_exprs.py#L88) 

We need to find a solution to this as without the ufuncs being rewritten as array expression we will not generate a parfor node and in turn will not create a GPU kernel for it.